velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp

```diff
@@ -135,21 +135,51 @@ class S3ReadFile final : public ReadFile {
 
 namespace filesystems {
 
-namespace S3Config {
-namespace {
-constexpr char const* kPathAccessStyle{"hive.s3.path-style-access"};
-constexpr char const* kEndpoint{"hive.s3.endpoint"};
-constexpr char const* kSecretKey{"hive.s3.aws-secret-key"};
-constexpr char const* kAccessKey{"hive.s3.aws-access-key"};
-constexpr char const* kSSLEnabled{"hive.s3.ssl.enabled"};
-constexpr char const* kUseInstanceCredentials{
-    "hive.s3.use-instance-credentials"};
-} // namespace
-} // namespace S3Config
+class S3Config {
+ public:
+  S3Config(const Config* config) : config_(config) {}
+
+  // Virtual addressing is used for AWS S3 and is the default (path-style-access
+  // is false). Path access style is used for some on-prem systems like Minio.
+  bool useVirtualAddressing() const {
+    return !config_->get("hive.s3.path-style-access", false);
+  }
+
+  bool useSSL() const {
+    return config_->get("hive.s3.ssl.enabled", true);
+  }
+
+  bool useInstanceCredentials() const {
+    return config_->get("hive.s3.use-instance-credentials", false);
+  }
+
+  bool hasAccessKey() const {
+    return config_->isValueExists("hive.s3.aws-access-key");
+  }
+
+  bool hasSecretKey() const {
+    return config_->isValueExists("hive.s3.aws-secret-key");
+  }
+
+  std::string endpoint() const {
+    return config_->get("hive.s3.endpoint", std::string(""));
+  }
+
+  std::string accessKey() const {
```

Comment:
Would it make sense to return std::optional for the properties which do not have a default value? Otherwise, how would the user know that they have to call hasAccessKey before calling this method? Would it make sense to group hasXxx() with the corresponding xxx() method?

Reply:
I've been replacing all instance of folly::Optional with std::optional in the code base. It would be nice to avoid introducing new instances.